### Parameter Defaults - Beware!

Řešíme problém, když máme parametry které se načtou do paměti a při volání je nepřegenrují ale pouze zavolají. Typickí příklad je s časem, kdy máme stále stejný čas. Proto je vhodné dávat defaultně None hodnoty pro tyto druhy parametrů a pokud nejsou non tak je volat.

In [85]:
from datetime import datetime

In [88]:
print(datetime.utcnow())

2020-05-05 17:13:00.797579


In [90]:
print(datetime.utcnow()) #vidíme, že čas se mění

2020-05-05 17:13:16.059331


In [97]:
def log(msg, *, dt=datetime.utcnow()):
    print(f"{dt} and {msg}")

In [98]:
log("message 1", dt="2001-01-01 ...")

2001-01-01 ... and message 1


Jasná funkce a něco plodíme, čas jsem si zadal svůj takže je vše ok.

In [99]:
log("message 2")

2020-05-05 17:14:55.640335 and message 2


Nyní vidíme, že když nezadáme datetime použijeme funkci.

In [100]:
log("message 3")

2020-05-05 17:14:55.640335 and message 3


#### A TO JE NÁŠ PROBLÉM - ČAS SE NEZMĚNIL

2020-05-05 17:14:55.64033, 2020-05-05 17:14:55.64033

Defaultní hodnota totiž byla vytvořena v čtverci IN[97] - pokud bych jí znovu spustil a přehrál funkce, mám zase 2 stejné časy, jen jiné.

Zkusíme to tedy vyřešit skrze None hodnotu

In [102]:
def log(msg, *, dt=None):
    if not dt:
    #pokud dt je falsy - pokud jsem jej nezadal tak načtu dle funkce
        dt = datetime.utcnow()
    print(f"{dt} and {msg}")

In [105]:
log("message 4")

2020-05-05 17:18:32.764441 and message 4


In [106]:
log("message 4")

2020-05-05 17:18:33.354833 and message 4


A nyní krásně vidím, že mám vše v pořádku :). Ale lepší zápis.

In [108]:
def log(msg, *, dt=None):
    dt = dt or datetime.utcnow()
    # pokud dt má hodnotu použije dt 
    # nebo použije datetime.utcnow() :)
    print(f"{dt} and {msg}")

Teď trochu z jiného soudku

In [109]:
my_list = [1,2,3]

def func(a=my_list):
    print(a)

In [110]:
func()

[1, 2, 3]


In [111]:
func(["a","b"])

['a', 'b']


In [112]:
my_list.append(4)

In [113]:
my_list

[1, 2, 3, 4]

In [114]:
func()

[1, 2, 3, 4]


Jaký je zde problém? Mám ve funkci odkaz na objekt mimo. Pokud tedy nechci aby nedošlo ke změně je lepší použít tuple.